swift/tools/identify-environment.sh

```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+set -eu
+
+echo '{ "configuration": { "os": "macos-latest", "swift": {} }}'
```

Comment:
`macos-latest` is the name of a GitHub hosted Actions runner, not the name of an OS. Does it make sense to return this value regardless of the environment in which this script is run?

Reply:
It is the name of a runner image and that's convenient for direct applications in GHA workflows. For other applications, I don't see a strong reason to invent a different naming scheme since the runner image format is `<name>-<version>` so it is always possible to break them up into the OS name and the OS version if needed. We could do this ourselves and provide one field for the OS name and one for the version, but then we'd always have to stitch them back together in GHA workflows.

Reply:
I can see why this is convenient. However, I disagree with using the name `os` when in fact you mean "action runner name" and not operation system. If you use the name `os` for this then it will be hard to add more OS-related information in the future in a backwards compatible way. You can't go from `{ "os": "macos-latest}` to `{ "os": { "runner": "macos-latest", "name": "MacOS", "version": "12.3.1"}}` because clients that expect `os` to be a string would fail if it is suddenly an object. A common trick to ensure backwards compatibility is to use objects and make clients resilient against unknown fields.

Reply:
I agree with your general point about using JSON objects to ensure backwards compatibility -- that's why we already don't have the `"configuration"` object at the top-level, but in an object all by itself. At deeper levels, it's obviously not as clear cut because ultimately we do need to have values other than objects at some point. 

It seems that there are two parts to your comment:

- `"os"` is a bad name for an OS + version string.
- There is potentially information, other than an OS name and a version, relating to an operating system choice that could become useful in the future.

I have thought about this a bit over the last day and I think I am happy to change `"os"` into an object containing separate fields for the OS name and version. In the bigger picture, the main piece of information that we need elsewhere is a runner image name, but the new action that we are working on can infer that from the OS name and version and then inject it into the JSON. I think that makes sense since that's actually Actions-specific.

Reply:
I would definitely stay away from implementing this in the language-specific scripts. If we ever want to choose something other than `<os>-latest`, we'd need to potentially duplicate a bunch of logic around which GHA image versions actually exist (which is in itself a problem with the other two approaches as well).

Implementing it in the CLI removes the potential need for duplication of that logic, but the CLI is not GHA-specific. If it makes a recommendation for a GHA image, would customers then also expect us to make recommendations for other CI systems? 

The Action is the one GHA-specific component, so it makes sense for it to add the GHA-specific information to the output. 

For each of the options, we will need to be mindful of how we handle the process of recommending a runner image. E.g. if we had `{ "name": "macOS", "version": "12" }` and recommended `macos-12`, then that would have been fine a while ago, but that image is no longer available now since they have been replaced by the `macos-13` image.

Reply:
> I would definitely stay away from implementing this in the language-specific scripts. If we ever want to choose something other than `<os>-latest`, we'd need to potentially duplicate a bunch of logic around which GHA image versions actually exist (which is in itself a problem with the other two approaches as well).

Is there more duplication here than languages defining a fallback of `macos-latest` if they really wanted `macos-12`?  I agree the CLI might be a better place, but would like to double check the extent of this.

> Implementing it in the CLI removes the potential need for duplication of that logic, but the CLI is not GHA-specific. If it makes a recommendation for a GHA image, would customers then also expect us to make recommendations for other CI systems?
> 
> The Action is the one GHA-specific component, so it makes sense for it to add the GHA-specific information to the output.

The CLI does not only run on Actions, but that doesn't mean it shouldn't be knowledgeable about Actions if that can produce a better experience.  For example, the CLI can produce tailored help when it fails depending whether it's running on Actions or not.  I think here we have a case for a better experience being `codeql resolve build-environment` producing the same output as the Action.
 
> For each of the options, we will need to be mindful of how we handle the process of recommending a runner image. E.g. if we had `{ "name": "macOS", "version": "12" }` and recommended `macos-12`, then that would have been fine a while ago, but that image is no longer available now since they have been replaced by the `macos-13` image.

I recommend we defer this decision until we start thinking about multiple configurations.  For installing the right Go version in default setup, I don't think we need to support this.  If we're not happy with the design, we could even drop this and have Go just produce the version of Go to install.

Reply:
> I recommend we defer this decision until we start thinking about multiple configurations.

Agreed. For now, we don't need to worry about this.